### PR TITLE
Reduce the number of recalc styles during a typically AMP boot procedure from 4 to 2

### PR DESCRIPTION
**Pull request by cramforce**
which is likely the minimum possible right now.

Refactors both viewport and resources to use the service pattern
instead of being magic global singletons.

### CODE REVIEW

**Comment by cramforce** [comment]
CC @paul_irish

File: src/viewport.js [hunk 1]
```diff
@@ -123,6 +123,8 @@ export class Viewport {
    * @return {number}
    */
   getScrollTop() {
+    if (!this./*OK*/scrollTop_)
```
Comment (dvoytenko) [inline src/viewport.js]: (this.scrollTop_ === null)?

Otherwise will keep requesting for 0.

File: src/viewport.js [hunk 2]
```diff
@@ -131,6 +133,8 @@ export class Viewport {
    * @return {number}
    */
   getScrollLeft() {
+    if (!this./*OK*/scrollleft_)
```
Comment (dvoytenko) [inline src/viewport.js]: Ditto with `null`

File: src/viewport.js [hunk 3]
```diff
@@ -139,7 +143,14 @@ export class Viewport {
    * @return {!{width: number, height: number}}
    */
   getSize() {
-    return this.binding_.getSize();
+    var size = this.binding_.getSize();
```
Comment (dvoytenko) [inline src/viewport.js]: I'd much rather cache the whole "size" object and get rid of `this.width_` completely. And this method should cache the value as the case with others.

Comment (dvoytenko) [inline src/viewport.js]: That way the line will be clear: Viewport does caching and Binding does not - just translates API directly.

File: src/viewport.js (hunk 1 above)
Comment (cramforce) [inline src/viewport.js]: Done.

File: src/viewport.js (hunk 2 above)
Comment (cramforce) [inline src/viewport.js]: Done

File: src/viewport.js (hunk 3 above)
Comment (cramforce) [inline src/viewport.js]: Done.

**Comment by cramforce** [comment]
PTAL

File: src/viewport.js [hunk 4]
```diff
@@ -133,6 +138,8 @@ export class Viewport {
    * @return {number}
    */
   getScrollTop() {
+    if (this./*OK*/scrollTop_ == null)
```
Comment (dvoytenko) [inline src/viewport.js]: Here and below, `{}` style guide for `if`.

File: src/viewport.js [hunk 5]
```diff
@@ -366,9 +378,10 @@ export class Viewport {
 
   /** @private */
   resize_() {
-    let oldWidth = this.width_;
-    this.width_ = this.getSize().width;
-    this.changed_(oldWidth != this.width_, 0);
+    let oldSize = this.size_;
+    this.size_ = null;  // Need to recalc.
```
Comment (dvoytenko) [inline src/viewport.js]: Why not immediately do `this.size_ = this.getSize()` since it's read anyway?

File: src/viewport.js (hunk 4 above)
Comment (erwinmombay) [inline src/viewport.js]: i'll add it to the linter

Comment (erwinmombay) [inline src/viewport.js]: nvm, its already there `[02:10:02] /home/travis/build/ampproject/amphtml/src/viewport.js: line 141, col 5, Error - Expected { after 'if' condition. (curly)`

Comment (cramforce) [inline src/viewport.js]: Wow, I'm not usually this sloppy :)

File: src/viewport.js (hunk 5 above)
Comment (cramforce) [inline src/viewport.js]: I'm not sure I follow. getSize is cached now, so one needs to to bust the cache before reading when it is known that the size changed.

Comment (dvoytenko) [inline src/viewport.js]: Got it. Missed it. I think doing `this.size_ = this.binding_.getSize();` here would be a bit safer. But I don't feel to strongly about this.

**Comment by dvoytenko** [comment]
LGTM. Not sure what's in Travis.

**Comment by erwinmombay** [comment]
@cramforce can you add a skip to that carousel test, will investigate and fix later.

**Pull request merged**